### PR TITLE
(Please Merge) Merge Zhenhui_lgb into Staging

### PR DESCRIPTION
**Pull request by motefly**
### Description

Add LightGBM application for Recommenders on a sample data from CRITEO, please refer to "notebooks/00_quick_start/lightgbm_tinycriteo.ipynb".
Additionally, there are some utils file added to "reco_utils/recommender/lightgbm" and related package like "lightgbm==2.2.1" and "category_encoders>=1.3.0".
Pass the smoke test with "pytest tests/smoke/test_notebooks_python.py::test_lightgbm_quickstart_smoke".

### Related Issues

None.

### Checklist:

- [x] My code follows the code style of this project, as detailed in our [contribution guidelines](../CONTRIBUTING.md).
- [x] I have added tests.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: reco_utils/recommender/lightgbm/lightgbm_utils.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    remote_path = azure_container_url + remote_resource_name
+    maybe_download(remote_path, remote_resource_name, data_path)
+    tar_ref = tarfile.open(os.path.join(data_path, remote_resource_name), "r")
+    tar_ref.extractall(data_path)
+    tar_ref.close()
+    os.remove(os.path.join(data_path, remote_resource_name))
+
+def cal_metric(labels, preds, metrics):
+    """Calculate metrics,such as auc, logloss
+    FIXME: refactor this with the reco metrics
```
Comment (yueguoguo) [inline reco_utils/recommender/lightgbm/lightgbm_utils.py]: Just notice the FIXME - yea it would be better to use the reco metrics consistently.

File: reco_utils/recommender/lightgbm/lightgbm_utils.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        self.save_value_filter = {}
+        self.save_num_embs = {}
+        self.Max_len = {}
+        self.samples = 0
+
+    def fit_transform(self, inPath):
+        logging.info('----------------------------------------------------------------------')
+        logging.info('Fitting and Transforming %s .'%inPath)
+        logging.info('----------------------------------------------------------------------')
+        df = pd.read_csv(inPath, dtype=self.dtype_dict)
```
Comment (yueguoguo) [inline reco_utils/recommender/lightgbm/lightgbm_utils.py]: Do we want to accept a dataframe instead of a file path as input, which might make it easier for users that interface with a prepared data for model fitting and transforming?

File: reco_utils/recommender/lightgbm/lightgbm_utils.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+            del feats
+            gc.collect()
+        trn_y = np.array(df[self.label_name].values).reshape((-1,1))
+        del df
+        gc.collect()
+        trn_x = np.array(rows)
+        return (trn_x, trn_y)
+
+    # for test dataset
+    def transform(self, inPath):
```
Comment (yueguoguo) [inline reco_utils/recommender/lightgbm/lightgbm_utils.py]: Same here as above - it would be great to allow dataframe as input

File: notebooks/00_quick_start/lightgbm_tinycriteo.ipynb [hunk 4]
```diff
@@ -0,0 +1,827 @@
...
+    "<i>Copyright (c) Microsoft Corporation. All rights reserved.</i>\n",
+    "\n",
+    "<i>Licensed under the MIT License.</i>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<i>Copyright (c) Microsoft Corporation. All rights reserved.</i>\n",
```
Comment (yueguoguo) [inline notebooks/00_quick_start/lightgbm_tinycriteo.ipynb]: Remove the duplicated heading

File: reco_utils/recommender/lightgbm/lightgbm_utils.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+from tqdm import tqdm
+import collections
+import gc
+
+import tarfile
+from reco_utils.dataset.url_utils import maybe_download
+
+logging.basicConfig(level = logging.INFO, format = '%(asctime)s [INFO] %(message)s')
+
+def download_lgb_resources(azure_container_url, data_path, remote_resource_name):
```
Comment (gramhagen) [inline reco_utils/recommender/lightgbm/lightgbm_utils.py]: would be better to consolidate dataset loading under reco_utils/datasets

File: reco_utils/recommender/lightgbm/lightgbm_utils.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+            raise ValueError("not define this metric {0}".format(metric))
+    return res
+
+def unpackbits(x,num_bits):
+    xshape = list(x.shape)
+    x = x.reshape([-1,1])
+    to_and = 2**np.arange(num_bits).reshape([1,num_bits])
+    return (x & to_and).astype(bool).astype(int).reshape(xshape + [num_bits])
+
+class NumEncoder(object):
```
Comment (gramhagen) [inline reco_utils/recommender/lightgbm/lightgbm_utils.py]: can you provide some detail on what this class is doing and how it is used. I like that you're leveraging the categorical_encoder package, it's really nice. just curious if this functionality is something that we can leverage elsewhere, but it's a little hard to follow what's happening.

File: reco_utils/recommender/lightgbm/lightgbm_utils.py (hunk 5 above)
Comment (yueguoguo) [reply reco_utils/recommender/lightgbm/lightgbm_utils.py]: +1

File: reco_utils/recommender/lightgbm/lightgbm_utils.py (hunk 6 above)
Comment (motefly) [reply reco_utils/recommender/lightgbm/lightgbm_utils.py]: Well, the steps of NumEncoder was shown in the jupyter notebook. For much clear showing, I'll add some necessary comments in the code.

File: reco_utils/recommender/lightgbm/lightgbm_utils.py (hunk 1 above)
Comment (motefly) [reply reco_utils/recommender/lightgbm/lightgbm_utils.py]: I find some differences when dealing with the metric calculations, that there is no any user columns or item columns in CRITEO data. So I cannot reuse the function defined like https://github.com/Microsoft/Recommenders/blob/zhenhui_lgb/reco_utils/evaluation/python_evaluation.py#L603. The single call for sklearn.metric.auc is all I need.
So what I should change for this comment?

File: reco_utils/recommender/lightgbm/lightgbm_utils.py (hunk 5 above)
Comment (motefly) [reply reco_utils/recommender/lightgbm/lightgbm_utils.py]: Is it means that I'd better add a new file named "criteo.py" into that dataset directory?

File: reco_utils/recommender/lightgbm/lightgbm_utils.py (hunk 3 above)
Comment (motefly) [reply reco_utils/recommender/lightgbm/lightgbm_utils.py]: Fixed.

File: notebooks/00_quick_start/lightgbm_tinycriteo.ipynb (hunk 4 above)
Comment (motefly) [reply notebooks/00_quick_start/lightgbm_tinycriteo.ipynb]: Fixed.

File: reco_utils/recommender/lightgbm/lightgbm_utils.py (hunk 2 above)
Comment (motefly) [reply reco_utils/recommender/lightgbm/lightgbm_utils.py]: Fixed.

File: reco_utils/recommender/lightgbm/lightgbm_utils.py (hunk 1 above)
Comment (yueguoguo) [reply reco_utils/recommender/lightgbm/lightgbm_utils.py]: I see. Let's keep the comment and later fix it. The AUC/logloss evaluation can be amended such that it allows data in a different schema to be evaluated.

File: reco_utils/recommender/lightgbm/lightgbm_utils.py (hunk 5 above)
Comment (miguelgfierro) [reply reco_utils/recommender/lightgbm/lightgbm_utils.py]: Agree with this. in the `criteo.py` we can add something like we have in movielens: https://github.com/Microsoft/Recommenders/blob/master/reco_utils/dataset/movielens.py#L133

I think it will be good to take a look at what @jreynolds01 is doing: he has a function called load_spark_df https://github.com/Microsoft/Recommenders/blob/jeremr-criteo-dataloader/reco_utils/dataset/criteo_dac.py#L23, we can do a function called `load_pandas_df`.

Ideally, we want to do single responsibility, like we were discussing here: https://github.com/Microsoft/Recommenders/pull/575#discussion_r260704511

Also, @motefly, we have a function called `maybe_download` that can be useful to you  https://github.com/Microsoft/Recommenders/blob/master/reco_utils/dataset/url_utils.py#L11

File: tests/smoke/test_notebooks_python.py [hunk 7]
```diff
@@ -91,3 +91,26 @@ def test_vw_deep_dive_smoke(notebooks):
     assert results["ndcg"] == pytest.approx(0.65339, rel=TOL, abs=ABS_TOL)
     assert results["precision"] == pytest.approx(0.514738, rel=TOL, abs=ABS_TOL)
     assert results["recall"] == pytest.approx(0.25684, rel=TOL, abs=ABS_TOL)
+
```
Comment (miguelgfierro) [inline tests/smoke/test_notebooks_python.py]: hey @motefly, do you know how long does this take?

File: tests/conftest.py [hunk 8]
```diff
@@ -159,6 +159,9 @@ def notebooks():
         "dkn_quickstart": os.path.join(
             folder_notebooks, "00_quick_start", "dkn_synthetic.ipynb"
         ),
+        "lightgbm_quickstart": os.path.join(
```
Comment (miguelgfierro) [inline tests/conftest.py]: apart from the smoke test that you added, we should add a unit test here https://github.com/Microsoft/Recommenders/blob/master/tests/unit/test_notebooks_python.py, this just runs the notebook with as fast as possible. 

In other examples you will see that we are just runing the notebooks, but for algos that take too long, we restrict the number of epochs like in this one https://github.com/Microsoft/Recommenders/blob/master/tests/unit/test_notebooks_gpu.py#L38. This is because we don't want the unit tests to take too long

File: tests/smoke/test_notebooks_python.py (hunk 7 above)
Comment (motefly) [reply tests/smoke/test_notebooks_python.py]: For the lightgbm_quickstart, it takes about 60 seconds.

Comment (miguelgfierro) [reply tests/smoke/test_notebooks_python.py]: awesome, thanks

File: notebooks/00_quick_start/lightgbm_tinycriteo.ipynb [hunk 9]
```diff
@@ -0,0 +1,802 @@
+{
```
Comment (gramhagen) [inline notebooks/00_quick_start/lightgbm_tinycriteo.ipynb]: <p>This notebook will give you a quick example of how to train LightGBM model in recommendation scenario.</p><p><br></p><p>to</p><p><br></p><p>This notebook will give you a quick example of how to train LightGBM model in a recommendation scenario.</p>

 _Reply via <a href='https://app.reviewnb.com/Microsoft/Recommenders/pull/611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (gramhagen) [inline notebooks/00_quick_start/lightgbm_tinycriteo.ipynb]: <p>Remove: 'After data preparation' as this now preceeds that section</p><p><br></p><p>This is predicting click or no click, might be helpful to state that directly for clarity.</p><p><br></p><p>Also, helpful to clarify that Area Under the Curve (AUC) is used as a metric which is less effected by imbalance in the classes of the dataset.</p>

 _Reply via <a href='https://app.reviewnb.com/Microsoft/Recommenders/pull/611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (gramhagen) [inline notebooks/00_quick_start/lightgbm_tinycriteo.ipynb]: <p>is the plan to update this based on the dataset loader?</p>

 _Reply via <a href='https://app.reviewnb.com/Microsoft/Recommenders/pull/611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (gramhagen) [inline notebooks/00_quick_start/lightgbm_tinycriteo.ipynb]: <p>better to use a temporary directory in case data_path does not exist</p><p><br></p><p>example using context manager:</p><p>```</p><p>from tempfile import TemporaryDirectory</p><p><strong>with</strong> TemporaryDirectory() <strong>as</strong> tmp:</p><p>    save_file = os.path.join(tmp, 'finished.model')</p><p>    lgb_model.save_model(save_file)</p><p>    ....</p><p>```</p>

 _Reply via <a href='https://app.reviewnb.com/Microsoft/Recommenders/pull/611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Comment (motefly) [reply notebooks/00_quick_start/lightgbm_tinycriteo.ipynb]: Fixed.

Comment (motefly) [reply notebooks/00_quick_start/lightgbm_tinycriteo.ipynb]: fixed, thanks.

Comment (motefly) [reply notebooks/00_quick_start/lightgbm_tinycriteo.ipynb]: Fixed, thanks.

File: reco_utils/recommender/lightgbm/lightgbm_utils.py (hunk 5 above)
Comment (motefly) [reply reco_utils/recommender/lightgbm/lightgbm_utils.py]: I have added the `criteo_dac.py` and the `load_pandas_df` function. Please review it again.

File: notebooks/00_quick_start/lightgbm_tinycriteo.ipynb (hunk 9 above)
Comment (motefly) [reply notebooks/00_quick_start/lightgbm_tinycriteo.ipynb]: I am sorry I don't know this comment refers to which line, and I added a data loader for CRITEO sample in `reco_utils/dataset/criteo_dac.py`, please review it. Thanks.

File: tests/conftest.py (hunk 8 above)
Comment (motefly) [reply tests/conftest.py]: I added a unit test for lightgbm, and it is passed. 
However, when I test the whole `tests/unit/test_notebooks_python.py`, it meets failure on the function `test_vw_deep_dive_runs`, both in my branch and master, and the `02_model/vowpal_wabbit_deep_dive.ipynb` also raises exception in Cell "In [7]". I am so confused about that...

Comment (miguelgfierro) [reply tests/conftest.py]: mmm that's weird, can you please paste the error? also, are you on windows or linux?

**Comment by miguelgfierro** [comment]
hey @motefly, just a minor ask, Le added a new utility function for auc and logloss https://github.com/Microsoft/Recommenders/blob/staging/reco_utils/evaluation/python_evaluation.py#L603, would you mind to use these functions? it has some checks on top of the usual sklearn ones

File: reco_utils/recommender/lightgbm/lightgbm_utils.py (hunk 5 above)
Comment (miguelgfierro) [reply reco_utils/recommender/lightgbm/lightgbm_utils.py]: this is great. I'm working with @jreynolds01 on the data loader for spark https://github.com/Microsoft/Recommenders/pull/619, when your code is merged I will modify it a little to be able to use both the small and the big criteo

File: tests/conftest.py (hunk 8 above)
Comment (motefly) [reply tests/conftest.py]:
```
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
<ipython-input-7-b44d564263fc> in <module>
     17                 test_params=test_params,
     18                 test_data=test,
---> 19                 prediction_path=prediction_path)
     20 
     21 comparison = pd.DataFrame(result, index=['Linear Regression'])

<ipython-input-3-3fbf83133880> in run_vw(train_params, test_params, test_data, prediction_path, logistic)
     13     # train model
     14     train_start = process_time()
---> 15     run(train_params.split(' '), check=True)
     16     train_stop = process_time()
     17 

~/.conda/envs/reco_base/lib/python3.6/subprocess.py in run(input, timeout, check, *popenargs, **kwargs)
    421         kwargs['stdin'] = PIPE
    422 
--> 423     with Popen(*popenargs, **kwargs) as process:
    424         try:
    425             stdout, stderr = process.communicate(input, timeout=timeout)

~/.conda/envs/reco_base/lib/python3.6/subprocess.py in __init__(self, args, bufsize, executable, stdin, stdout, stderr, preexec_fn, close_fds, shell, cwd, env, universal_newlines, startupinfo, creationflags, restore_signals, start_new_session, pass_fds, encoding, errors)
    727                                 c2pread, c2pwrite,
    728                                 errread, errwrite,
--> 729                                 restore_signals, start_new_session)
    730         except:
    731             # Cleanup if the child failed starting.

~/.conda/envs/reco_base/lib/python3.6/subprocess.py in _execute_child(self, args, executable, preexec_fn, close_fds, pass_fds, cwd, env, startupinfo, creationflags, shell, p2cread, p2cwrite, c2pread, c2pwrite, errread, errwrite, restore_signals, start_new_session)
   1362                         if errno_num == errno.ENOENT:
   1363                             err_msg += ': ' + repr(err_filename)
-> 1364                     raise child_exception_type(errno_num, err_msg, err_filename)
   1365                 raise child_exception_type(err_msg)
   1366 

FileNotFoundError: [Errno 2] No such file or directory: 'vw': 'vw'
```
These are found when running the `02_model/vowpal_wabbit_deep_dive.ipynb` (similarly when unit testing) and I am on Linux.

Comment (gramhagen) [reply tests/conftest.py]: yeah, we need to skip these tests if vowpal wabbit is not installed, it's trying to run vw from subprocess and failing

**Comment by miguelgfierro** [comment]
I'll fix the error you are getting:
`ModuleNotFoundError: No module named 'lightgbm'`

File: tests/conftest.py (hunk 8 above)
Comment (motefly) [reply tests/conftest.py]: OK, now the unit testing of my commit is not successful, can I re-test it and how?

**Comment by motefly** [comment]
For https://github.com/Microsoft/Recommenders/pull/611#issuecomment-470539819 , I have added it into the conda generator, why does it happen? 
And the `auc` and `logloss`  @miguelgfierro  mentioned are still not compatiable with ours, that CIRTEO data have no ``item`` column and ``user`` column.

**Comment by yueguoguo** [comment]
> For #611 (comment) , I have added it into the conda generator, why does it happen?
And the auc and logloss @miguelgfierro mentioned are still not compatiable with ours, that CIRTEO data have no item column and user column.

@miguelgfierro let's leave the codes as they are now, as the reco util for calculating auc and logloss requires the input dataframes to have user/item column, which is not present in the criteo dataset.

**Comment by miguelgfierro** [comment]
> And the auc and logloss @miguelgfierro mentioned are still not compatiable with ours, that CIRTEO data have no item column and user column.
oh yeah,  that makes sense. 

@motefly the test error was because, as of now we are not installing the libraries every time we are doing a PR. We will change this behavior in the future. I just installed the libraries manually and you shouldn't see this error again. However I see a different error now:
```
                                   [ 12%]
tests/unit/test_notebooks_python.py ......F                              [100%]

=================================== FAILURES ===================================
________________________________ test_lightgbm _________________________________

notebooks = {'als_deep_dive': '/data/home/recocat/cicd/3/s/notebooks/02_model/als_deep_dive.ipynb', 'als_pyspark': '/data/home/rec...baseline_deep_dive.ipynb', 'data_split': '/data/home/recocat/cicd/3/s/notebooks/01_prepare_data/data_split.ipynb', ...}

    @pytest.mark.notebooks
    def test_lightgbm(notebooks):
        notebook_path = notebooks["lightgbm_quickstart"]
        pm.execute_notebook(notebook_path, OUTPUT_NOTEBOOK, kernel_name=KERNEL_NAME,
                            parameters=dict(MAX_LEAF = 32,
                                            MIN_DATA = 20,
                                            NUM_OF_TREES = 10,
                                            TREE_LEARNING_RATE = 0.15,
                                            EARLY_STOPPING_ROUNDS = 20,
>                                           METRIC = "auc"),)

tests/unit/test_notebooks_python.py:63: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/anaconda/envs/reco_base/lib/python3.6/site-packages/papermill/execute.py:94: in execute_notebook
    raise_for_execution_errors(nb, output_path)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

nb = {'cells': [{'cell_type': 'code', 'metadata': {'inputHidden': True, 'hide_input': True}, 'execution_count': None, 'sour...end_time': '2019-03-07T14:57:50.636789', 'duration': 3.432436, 'exception': True}}, 'nbformat': 4, 'nbformat_minor': 2}
output_path = 'output.ipynb'

    def raise_for_execution_errors(nb, output_path):
        """Assigned parameters into the appropriate place in the input notebook
    
        Parameters
        ----------
        nb : NotebookNode
           Executable notebook object
        output_path : str
           Path to write executed notebook
        """
        error = None
        for cell in nb.cells:
            if cell.get("outputs") is None:
                continue
    
            for output in cell.outputs:
                if output.output_type == "error":
                    error = PapermillExecutionError(
                        exec_count=cell.execution_count,
                        source=cell.source,
                        ename=output.ename,
                        evalue=output.evalue,
                        traceback=output.traceback,
                    )
                    break
    
        if error:
            # Write notebook back out with the Error Message at the top of the Notebook.
            error_msg = ERROR_MESSAGE_TEMPLATE % str(error.exec_count)
            error_msg_cell = nbformat.v4.new_code_cell(
                source="%%html\n" + error_msg,
                outputs=[
                    nbformat.v4.new_output(output_type="display_data", data={"text/html": error_msg})
                ],
                metadata={"inputHidden": True, "hide_input": True},
            )
            nb.cells = [error_msg_cell] + nb.cells
            write_ipynb(nb, output_path)
>           raise error
E           papermill.exceptions.PapermillExecutionError: 
E           ---------------------------------------------------------------------------
E           Exception encountered at "In [5]":
E           ---------------------------------------------------------------------------
E           FileNotFoundError                         Traceback (most recent call last)
E           <ipython-input-5-b0eb7a9ffa83> in <module>
E                 5 label_col = 'Label'
E                 6 
E           ----> 7 all_data = criteo_dac.load_pandas_df(file_path, label_col, nume_cols, cate_cols)
E                 8 # split data to 3 sets
E                 9 length = len(all_data)
E           
E           /data/home/recocat/cicd/3/s/reco_utils/dataset/criteo_dac.py in load_pandas_df(local_cache_path, label_col, nume_cols, cate_cols)
E                34 
E                35   ## download and untar the data file
E           ---> 36   datapath = _load_datafile(local_cache_path=local_cache_path)
E                37   df = pd.read_csv(datapath, sep="\t", header=None, names=[label_col]+nume_cols+cate_cols)
E                38 
E           
E           /data/home/recocat/cicd/3/s/reco_utils/dataset/criteo_dac.py in _load_datafile(local_cache_path)
E                49           "http://labs.criteo.com/wp-content/uploads/2015/04/dac_sample.tar.gz",
E                50           filename,
E           ---> 51           work_directory=path,
E                52     )
E                53 
E           
E           /data/home/recocat/cicd/3/s/reco_utils/dataset/url_utils.py in maybe_download(url, filename, work_directory, expected_bytes)
E                23     filepath = os.path.join(work_directory, filename)
E                24     if not os.path.exists(filepath):
E           ---> 25         filepath, _ = urlretrieve(url, filepath)
E                26     else:
E                27         log.debug("File {} already downloaded".format(filepath))
E           
E           /anaconda/envs/reco_base/lib/python3.6/urllib/request.py in urlretrieve(url, filename, reporthook, data)
E               256         # Handle temporary file setup.
E               257         if filename:
E           --> 258             tfp = open(filename, 'wb')
E               259         else:
E               260             tfp = tempfile.NamedTemporaryFile(delete=False)
E           
E           FileNotFoundError: [Errno 2] No such file or directory: '/data/home/recocat/cicd/tests/resources/lightgbm/dac_sample.tar.gz'

/anaconda/envs/reco_base/lib/python3.6/site-packages/papermill/execute.py:241: PapermillExecutionError

```

**Comment by motefly** [comment]
For https://github.com/Microsoft/Recommenders/pull/611#issuecomment-470560009 , I have fixed it, since the directory used in notebook is not made before, I replace it with a tmp dir now. Besides, all checks have passed.
